### ISSUE TITLE
[Feature Request] make OrtSession#native run public or refactor to avoid array allocation on every inference

### ISSUE THREAD

**Issue by brettfazio** [microsoft/onnxruntime#13574]
### Describe the feature request

The public OrtSession#run function allocates multiple arrays for each call to inference. Likely in most programs the size of these arrays are never going to change. So either this function should be refactored to avoid allocation on every single inference or expose the native run function ( https://github.com/microsoft/onnxruntime/blob/main/java/src/main/java/ai/onnxruntime/OrtSession.java#L421 ) as public so that the caller map opt to avoid allocation

### Describe scenario use case

Avoid allocation of arrays for equivalent inferences.

**Comment by Craigacp** [comment]
Is this actually causing a performance issue? The JVM is pretty good at reallocating recently discarded objects, and none of these arrays live for very long. I agree it'll need to zero them again, but that's very cheap.

**Comment by Craigacp** [comment]
One thing that is planned is to allow users to supply their own output tensors, which would remove a lot of allocations in the output path to create them. If you're in the situation where the inputs and outputs are always the same size then that should help a lot (as you can hold onto the ByteBuffer for the inputs and rewrite them). There will still be a small number of allocations in both the Java and JNI parts of the run method, but the large allocations would all be gone.

**Comment by Craigacp** [comment]
I agree they escape in terms of the escape analysis but the GC should catch them with a young gen collection in G1 easily and they are usually a cheap allocation. Not sure how ZGC deals with them, maybe the generational one will help. I guess that given the remaining allocations are native then the GC doesn't see them so you might not have enough pressure to trigger a young gen G1 collection.

> If there's some proposal you think would be good to make a sort of 'fast path' into inference I would be happy to make the change too & add relevant tests.

It depends how fast a fast path you actually want. There's one which does essentially no allocation on the way in by storing the names as UTF-8 byte arrays stored in a `ByteBuffer`, passing in four arrays (input long*, input char*, output char*, output OnnxValue) that I can write to, and then passing in a list of `OnnxTensor`. The Java code then unpacks the pointers into the input long array, then the native code unpacks the byte buffer pointers into the char* arrays. It would be ugly to use, and relatively hard to make safe (to the point where it might be better to actually have a different session construction factory that binds the names on construction or a separate subclass of OrtSession). What I don't want to do is expose any method that accepts `long[]` as a pointer array as it's far too easy to make that crash everything. ORT always needs to manage the pointers to make sure they are in fact OrtValue pointers within the right lifespan.

That will still allocate (at least) two objects per output value, as each `OnnxTensor` needs to contain a `TensorInfo`, at least until we add support for supplying output values. At the point that support is added then this would need to be revised.

There are also changes coming which touch this code to incorporate support for sparse tensors which I'd like to land before making any changes to the run code.

**Comment by brettfazio** [comment]
Is it possible to cache these arrays at the session object level and resize / reallocate as necessary? Or are there multi threaded run call implication there that I might be unaware of (I'm just calling run from a single thread).

**Comment by Craigacp** [comment]
> Is it possible to cache these arrays at the session object level and resize / reallocate as necessary? Or are there multi threaded run call implication there that I might be unaware of (I'm just calling run from a single thread).

Yes, the run API is designed to be used from multiple threads concurrently, so I don't want to do any internal caching in the session object. We could potentially cache the names as those are fixed and it will only be a subset of them, but as ORT adds support for optional inputs then that caching and management becomes more complex.
